sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java

```diff
@@ -1348,6 +1440,108 @@ private void downloadToFileCleanup(AsynchronousFileChannel channel, String fileP
+        try {
+            return withContext(this::getTagsWithResponse);
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    Mono<Response<Map<String, String>>> getTagsWithResponse(Context context) {
+        return this.azureBlobStorage.blobs().getTagsWithRestResponseAsync(null, null, null, null, snapshot,
+            null /* versionId */, context)
+            .map(response -> {
+                Map<String, String> tags = new HashMap<>();
+                for(BlobTag tag : response.getValue().getBlobTagSet()) {
+                    tags.put(tag.getKey(), tag.getValue());
+                }
+                return new SimpleResponse<>(response, tags);
+            });
+    }
+
+    /**
+     * Sets user defined tags. The specified tags in this method will replace existing tags. If old values must be
+     * preserved, they must be downloaded and included in the call to this method.
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * {@codesnippet com.azure.storage.blob.specialized.BlobAsyncClientBase.setTags#Map}
+     *
+     * <p>For more information, see the
+     * <a href="ADD DOCS LINK">Azure Docs</a></p>
+     *
+     * @param tags Tags to associate with the blob.
+     * @return A reactive response signaling completion.
+     */
+    public Mono<Void> setTags(Map<String, String> tags) {
+        return this.setTagsWithResponse(tags).flatMap(FluxUtil::toMono);
+    }
+
+    /**
+     * Sets user defined tags. The specified tags in this method will replace existing tags. If old values must be
+     * preserved, they must be downloaded and included in the call to this method.
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * {@codesnippet com.azure.storage.blob.specialized.BlobAsyncClientBase.setTagsWithResponse#Map}
+     *
+     * <p>For more information, see the
+     * <a href="ADD DOCS LINK">Azure Docs</a></p>
+     *
+     * @param tags Tags to associate with the blob.
+     * @return A reactive response signaling completion.
+     */
+    public Mono<Response<Void>> setTagsWithResponse(Map<String, String> tags) {
+        try {
+            return withContext(context -> setTagsWithResponse(tags, context));
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    Mono<Response<Void>> setTagsWithResponse(Map<String, String> tags, Context context) {
+        List<BlobTag> tagList = null;
+        if (tags != null) {
+            tagList = new ArrayList<>();
+            for (String tag : tags.keySet()) {
+                tagList.add(new BlobTag().setKey(tag).setValue(tags.get(tag)));
+            }
+        }
+        BlobTags t = new BlobTags().setBlobTagSet(tagList);
+        return this.azureBlobStorage.blobs().setTagsWithRestResponseAsync(null, null, null, snapshot,
+            null /* versionId */, null, null, null, t, context)
```

Comment:
The spec says you can set tags on a version. It's unclear if it bumps the version number. It is, however, not allowed on a snapshot.

Do you think we should throw if the client has a snapshot value so that customers don't think we're setting tags on their snapshot when we're really just ignoring it?

Reply:
then we should be passing versionId here (should be available by now).

I think we can let service do the validation for snapshot (in case they change their mind).